lib/streamlit/caching.py

```diff
@@ -602,19 +616,25 @@ def has_changes(self) -> bool:
         context = Context(dict(caller_frame.f_globals, **caller_frame.f_locals), {}, {})
         code = compile(program, filename, "exec")
 
-        code_hasher = CodeHasher("md5")
-        code_hasher.update(code, context)
-        LOGGER.debug("Hashing block in %i bytes.", code_hasher.size)
+        hasher = hashlib.new("md5")
+        update_hash(
+            code,
+            hasher=hasher,
+            context=context,
+            hash_reason=HashReason.CACHING_BLOCK,
+            hash_source=code,
+        )
 
-        key = code_hasher.hexdigest()
+        key = hasher.hexdigest()
         LOGGER.debug("Cache key: %s", key)
 
         try:
             value, _ = _read_from_cache(
```

Comment:
Do you have to handle the `CachedObjectWasMutatedError` here as well since you're no longer handling it inside `read_from_cache`?

Reply:
That's already happening in line 500, no?

Reply:
On line 500 you're calling `_read_from_cache` via `def wrapped_func` of `def cache`.

On line 613 you're calling `_read_from_cache` via `def has_changes` of `Class Cache`.

Separate code paths

Reply:
Oh, good point!

We need to make sure st.Cache (capital C) supports those warnings too.

If you were wondering, this wasn't caught by tests because the tests for that codepath are deactivated -- since that whole feature is also deactivated by not including it in __init__.py

Reply:
Anyway, fixed it.